### PR TITLE
Convert package to ESM

### PR DESCRIPTION
**Pull request by rekmarks**
Converts this package to an ESM module. Kudos to [this guide](https://tsmx.net/convert-existing-nodejs-project-from-commonjs-to-esm/) and this issue: https://github.com/TypeStrong/ts-node/issues/1997

In our organization, this is only possible for this package because it is not used anywhere else, and we rely on CommonJS extensively.

Changes in detail:
- Update `package.json` for ESM compatibility
- Replace all `require()` with `import` statements
- Add `.js` file extension to all file imports in TypeScript source files
- Update TypeScript, ESLint, and Jest configs for ESM compatibility and recommended best practices
  - TypeScript (`tsconfig.json`)
    - Update the `module`/`moduleResolution` fields to `Node16`.
    - I chose `Node16` over `NodeNext`, as the behavior of `NodeNext` may change in the future. See [here](https://www.typescriptlang.org/docs/handbook/modules/reference.html#node16-nodenext) for details.
  - ESLint (`.eslintrc.cjs`)
    - `parseOptions.sourceType = 'module'`
  - Jest (`jest.config.cjs`)
    - Update the `moduleNameMapper` to strip the `.js` from local file imports. See [here](https://kulshekhar.github.io/ts-jest/docs/guides/esm-support/) for details.
- Replace `ts-node` with `tsx`, which has better ESM compatibility.
  - Specifically, the former does not work with Node v20 at the moment: https://github.com/TypeStrong/ts-node/issues/1997
- Update `jest-it-up`
  - This is so that we can use its `--config` option (thank you @PeterYinusa! https://github.com/rbardini/jest-it-up/pull/12)  to target our new `.cjs` config file.

### CODE REVIEW

**Comment by rekmarks** [comment]
I've checked all of these, the authors seem legit:
@SocketSecurity ignore @inquirer/type@1.1.5
@SocketSecurity ignore mute-stream@1.0.0
@SocketSecurity ignore esbuild@0.18.20
@SocketSecurity ignore tsx@4.6.1

[`@inquirer/type`](https://www.npmjs.com/package/@inquirer/type) is not unpublished, contrary to what the Socket Security report says.

**Comment by rekmarks** [comment]
@SocketSecurity ignore @inquirer/type@1.1.5 mute-stream@1.0.0 esbuild@0.18.20 tsx@4.6.1

I've checked all of these, the authors seem legit.
[`@inquirer/type`](https://www.npmjs.com/package/@inquirer/type) is not unpublished, contrary to what the Socket Security report says.

File: package.json [hunk 1]
```diff
@@ -58,28 +60,29 @@
...
     "stdio-mock": "^1.2.0",
     "ts-jest": "^29.1.0",
-    "ts-node": "^10.7.0",
+    "tsx": "^4.6.1",
     "typescript": "~5.1.6"
   },
   "packageManager": "yarn@3.2.1",
   "engines": {
-    "node": ">=14.0.0"
+    "node": ">=16.0.0"
```
[the remark this answers is not in the document]
Comment (rekmarks) [reply package.json:76]: See #114

File: .eslintrc.cjs [hunk 2]
```diff
@@ -3,6 +3,20 @@ module.exports = {
 
   extends: ['@metamask/eslint-config', '@metamask/eslint-config-nodejs'],
 
+  env: {
+    // See parserOptions comment
+    es2022: true,
```
[the remark this answers is not in the document]
Comment (rekmarks) [reply .eslintrc.cjs]: Done in [86398ec](https://github.com/MetaMask/create-release-branch/pull/113/commits/86398ec3ceb43166ca06e162a7ae452a0aa2400f)

**Comment by rekmarks** [comment]
@SocketSecurity ignore tsx@4.6.2

**Comment by mcmire** [comment]
@rekmarks Sorry for the conflicts. We just merged a PR that had a wide array of changes. I want to get this PR in the next release though, so I'm happy to work on resolving these conflicts for you — just let me know.

**Comment by rekmarks** [comment]
@mcmire no, no: someone had to suffer here, and under the circumstances it should be me. I'll resolve the conflicts!

**Comment by rekmarks** [comment]
Alright, this is ready for re-review!

**Review by mcmire — APPROVED** [review]
Looks good!

**Pull request merged**